gr-azure-software-radio/lib/difi_source_cpp_impl.cc

```diff
@@ -467,135 +395,144 @@ namespace gr {
-          d_num_bytes_in_read_buffer -= num_bytes_to_copy;
-          size_gotten = pkt_size;
-        }
-        else if(pkt_size == d_num_bytes_in_read_buffer)
+        // select file descriptors that we want to monitor (socket client)
+        fd_set read_fds;
+        FD_ZERO(&read_fds);
+        FD_SET(d_client_socket,&read_fds);
+
+        struct timeval tv;
+        tv.tv_sec = 0;
+        tv.tv_usec = 100000;
+        int status = select(d_client_socket+1, &read_fds, NULL, NULL, &tv);
+        if (status < 0)
         {
-          d_packet_buffer_start_idx = 0;
-          d_num_bytes_in_read_buffer = 0;
-          size_gotten = pkt_size;
+          // get the error code from the file descriptor
+          int error_code;
+          socklen_t len = sizeof(error_code);
+
+          getsockopt(d_client_socket, SOL_SOCKET, SO_ERROR, &error_code, &len);
+          GR_LOG_ERROR(this->d_logger, "Could not select client file descriptor -  (error code: " + std::to_string(error_code) + ")");
+          throw std::runtime_error("Could not select client file descriptor -  (error code: " + std::to_string(error_code) + ")");
         }
-        else
+
+        // check if there is data in the socket buffer
+        if(FD_ISSET(d_client_socket, &read_fds))
         {
-          size_gotten = recv(d_client_socket, &d_read_buffer[0], d_read_buffer.size(),MSG_WAITALL);
+          int size_gotten = recv(d_client_socket,buf+num_received,num_to_read,MSG_WAITALL);
 
           // when monitoring the file descriptor using 'select' and the fd is set, there should be
           // bytes in the socket buffer. If the receive call returns 0 bytes then the connection has
           // been lost and the client needs to re-connect
           if(size_gotten <= 0)
           {
             reset_tcp_connection();
-            d_packet_buffer_start_idx = 0;
-            d_num_bytes_in_read_buffer = 0;
-            return -1;
+            return 0;
           }
 
-          d_packet_buffer_start_idx = num_bytes_to_copy;
-          num_bytes_to_copy = std::min(pkt_size-d_num_bytes_in_read_buffer,size_gotten);
-
-          std::copy_n(d_read_buffer.cbegin(), num_bytes_to_copy, d_packet_buffer.begin()+d_num_bytes_in_read_buffer);
-
-          int num_bytes_pending = pkt_size-d_packet_buffer_start_idx-size_gotten;
+          num_to_read = len-num_received;
+          num_received += size_gotten;
+        }
 
-          if(num_bytes_pending > 0)
-          {
-            d_packet_buffer_start_idx += (d_packet_buffer_start_idx+size_gotten); // copy offset for remaining bytes
-            d_num_bytes_in_read_buffer = 0;
-            return -1;
-          }
-          else if(num_bytes_pending < 0)
-          {
-            d_packet_buffer_start_idx = num_bytes_to_copy;
-            d_num_bytes_in_read_buffer = abs(num_bytes_pending);
-            size_gotten = pkt_size;
-          }
-          else
+        gettimeofday(&end,0);
+        long seconds = end.tv_sec - begin.tv_sec;
```

Comment:
can we use chrono timer isntead? little easier and more modern cpp approach,


#include <chrono>
using namespace std::chrono;
 
auto start = high_resolution_clock::now()